macros/src/lib.rs

```diff
@@ -64,54 +64,72 @@ pub fn derive_bundle(input: TokenStream) -> TokenStream {
+                                if !dedup.insert(ty) {
+                                    panic!("{} has multiple {} fields; each type must occur at most once!", stringify!(#ident), name);
+                                }
                             }
-                        }
 
-                        let mut tys = [#((mem::align_of::<#tys>(), TypeId::of::<#tys>())),*];
-                        tys.sort_unstable_by(|x, y| x.0.cmp(&y.0).reverse().then(x.1.cmp(&y.1)));
-                        let mut ids = [TypeId::of::<()>(); #n];
-                        for (id, info) in ids.iter_mut().zip(tys.iter()) {
-                            *id = info.1;
-                        }
-                        ids
-                    };
-                }
+                            let mut tys = [#((mem::align_of::<#tys>(), TypeId::of::<#tys>())),*];
+                            tys.sort_unstable_by(|x, y| x.0.cmp(&y.0).reverse().then(x.1.cmp(&y.1)));
+                            let mut ids = [TypeId::of::<()>(); #n];
+                            for (id, info) in ids.iter_mut().zip(tys.iter()) {
+                                *id = info.1;
+                            }
+                            ids
+                        };
+                    }
 
-                f(&*ELEMENTS)
-            }
+                    f(&*ELEMENTS)
+                }
 
-            fn static_type_info() -> Vec<::hecs::TypeInfo> {
-                let mut info = vec![#(::hecs::TypeInfo::of::<#tys>()),*];
-                info.sort_unstable();
-                info
-            }
+                fn static_type_info() -> Vec<::hecs::TypeInfo> {
+                    let mut info = vec![#(::hecs::TypeInfo::of::<#tys>()),*];
+                    info.sort_unstable();
+                    info
+                }
 
-            unsafe fn get(
-                mut f: impl FnMut(::hecs::TypeInfo) -> Option<std::ptr::NonNull<u8>>,
-            ) -> Result<Self, ::hecs::MissingComponent> {
-                #(
-                    let #fields = f(::hecs::TypeInfo::of::<#tys>())
-                            .ok_or_else(::hecs::MissingComponent::new::<#tys>)?
-                            .cast::<#tys>()
-                        .as_ptr();
-                )*
-                Ok(Self { #( #fields: #fields.read(), )* })
+                unsafe fn get(
+                    mut f: impl FnMut(::hecs::TypeInfo) -> Option<std::ptr::NonNull<u8>>,
+                ) -> Result<Self, ::hecs::MissingComponent> {
+                    #(
+                        let #fields = f(::hecs::TypeInfo::of::<#tys>())
+                                .ok_or_else(::hecs::MissingComponent::new::<#tys>)?
+                                .cast::<#tys>()
+                            .as_ptr();
+                    )*
+                    Ok(Self { #( #fields: #fields.read(), )* })
+                }
```

Comment:
Rather than having two code paths here, could we resolve the issue with type annotations?